packages/contracts/contracts/SmartInvoiceEscrow.sol

```diff
@@ -664,30 +664,232 @@ contract SmartInvoiceEscrow is
+            uint256 currentRuling = arbitrator.currentRuling(
+                disputeData.disputeId
+            );
+            (originalCost, totalCost) = _appealCost(_ruling, currentRuling);
+            _checkAppealPeriod(_ruling, currentRuling);
+        }
+
+        uint256 lastRoundIndex = rounds.length - 1;
+        Round storage lastRound = rounds[lastRoundIndex];
+
+        if (lastRound.hasPaid[_ruling]) revert AppealFeeAlreadyPaid();
+        uint256 already = lastRound.paidFees[_ruling];
+
+        uint256 remaining = totalCost > already ? (totalCost - already) : 0;
+        uint256 contribution = msg.value < remaining ? msg.value : remaining;
+
+        // Record contribution
+        lastRound.paidFees[_ruling] = already + contribution;
+        lastRound.contributions[msg.sender][_ruling] += contribution;
+        emit Contribution(
+            _localDisputeId,
+            lastRoundIndex,
+            _ruling,
+            msg.sender,
+            contribution
+        );
+
+        // If fully funded now, mark and add to pot
+        if (lastRound.paidFees[_ruling] >= totalCost) {
+            lastRound.hasPaid[_ruling] = true;
+            lastRound.fundedRulings.push(_ruling);
+            lastRound.feeRewards += lastRound.paidFees[_ruling];
+            emit RulingFunded(_localDisputeId, lastRoundIndex, _ruling);
+        }
+
+        // If two opposing rulings are fully funded, escalate: pay base fee and open next round
+        if (lastRound.fundedRulings.length == 2) {
+            rounds.push(); // next round
+            uint256 pay = originalCost;
+            if (lastRound.feeRewards >= pay) {
+                lastRound.feeRewards -= pay;
+            } else {
+                pay = lastRound.feeRewards;
+                lastRound.feeRewards = 0;
+            }
+            arbitrator.appeal{value: pay}(
+                disputeData.disputeId,
+                arbitratorExtraData
+            );
+        }
+
+        // Refund any excess
+        uint256 refund = msg.value > contribution
+            ? (msg.value - contribution)
+            : 0;
+        if (refund > 0) {
+            (bool s, ) = payable(msg.sender).call{value: refund}("");
+            s;
+        }
+
+        return lastRound.hasPaid[_ruling];
+    }
+
+    /// @notice Withdraw reimbursable fees and/or rewards from a specific round
+    function withdrawFeesAndRewards(
+        uint256 _localDisputeId,
+        address payable _contributor,
+        uint256 _round,
+        uint256 _ruling
+    ) public override nonReentrant returns (uint256 amount) {
+        if (_localDisputeId != 0) revert IncorrectDisputeId();
+        if (locked) revert Locked();
+        if (!disputeData.isRuled) revert DisputeNotRuled();
+        if (_round >= rounds.length) return 0;
+
+        Round storage r = rounds[_round];
+        amount = _previewWithdrawableAmount(
+            r,
+            _contributor,
+            _ruling,
+            disputeData.ruling
```

Comment:
As mentioned above, overwriting `disputeData` with each new dispute can cause problems with withdrawals, since the ruling for past disputes won't be obtained